### PR TITLE
geometry editing

### PR DESCRIPTION
**Pull request by 3nids**
rebased version of #331 

Does not work on multi geometry layer (geometry will be left unchanged)

Coming later:
* [x] adding points
* [x] disable tool for multi layer
* [x] better colors for toolbar
* [x] icons for attributes vs geom editing

TODO:
* [x] test with layer CRS != map CRS

https://youtu.be/v4xseL2elfE

## Test APK
Last test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-5589632c7a752b2caef00120ca7fb8a83953b4b1-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-5589632c7a752b2caef00120ca7fb8a83953b4b1-x86.apk)

### CODE REVIEW

**Comment by 3nids** [comment]
image

ping @m-kuhn

**Comment by 3nids** [comment]
and for the tests
image

I have no issue when compiling locally

File: test/test_vertexmodel.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      QCOMPARE( mModel->mCurrentIndex, 1 );
+      QCOMPARE( mModel->currentPoint().x(), 1.0 );
+    }
+
+    void transformTest()
+    {
+      MapSettings *mapSettings = new MapSettings();
+      mapSettings->setDestinationCrs( QgsCoordinateReferenceSystem::fromEpsgId( 21781 ) );
+      mModel->setMapSettings( mapSettings );
+      QCOMPARE( mModel->mapSettings()->destinationCrs().authid(), "EPSG:21781" );
```
Comment (m-kuhn) [inline test/test_vertexmodel.cpp]: `QStringLiteral( )` and the error should be gone

File: android/AndroidManifest.xml [hunk 2]
```diff
@@ -39,15 +39,15 @@
       <meta-data android:name="android.app.qt_libs_resource_id" android:resource="@array/qt_libs"/>
```
Comment (m-kuhn) [inline android/AndroidManifest.xml]: Changes in this file should be reverted

File: scripts/astyle.sh [hunk 3]
```diff
@@ -27,20 +27,28 @@ fi
...
---style=allman \
---preserve-date \
---indent-preprocessor \
---convert-tabs \
---indent=spaces=2 \
---indent-classes \
---indent-labels \
---indent-namespaces \
---indent-switches \
+MY_ARTISTIC_STYLE_OPTIONS=" \
```
Comment (m-kuhn) [inline scripts/astyle.sh]: Should we merge this file already?
If you are afraid of merge/rebase problems, let's wait ;)

Comment (3nids) [reply scripts/astyle.sh]: no, you take all or nothing ;)

File: android/AndroidManifest.xml (hunk 2 above)
Comment (3nids) [reply android/AndroidManifest.xml]: fixed

File: test/test_vertexmodel.cpp (hunk 1 above)
Comment (3nids) [reply test/test_vertexmodel.cpp]: fixed

File: src/core/featurelistextentcontroller.cpp [hunk 4]
```diff
@@ -18,10 +18,10 @@
 #include <qgsvectorlayer.h>
 #include <qgsgeometry.h>
 
-FeatureListExtentController::FeatureListExtentController( QObject* parent )
+FeatureListExtentController::FeatureListExtentController( QObject *parent )
   : QObject( parent )
 {
-  connect( this, &FeatureListExtentController::autoZoomChanged, this, &FeatureListExtentController::zoomToSelected );
+  connect( this, &FeatureListExtentController::autoZoomChanged, this, [ = ]() {this->zoomToSelected();} );
```
Comment (m-kuhn) [inline src/core/featurelistextentcontroller.cpp]: No need for the `this->` part

File: src/core/vertexmodel.h [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    enum EditingMode
+    {
+      NoEditing,
+      EditVertex,
+      AddVertex
+    };
+    Q_ENUM( EditingMode )
+
+
+    struct Centroid
```
Comment (m-kuhn) [inline src/core/vertexmodel.h]: Can this be put into the private section?

File: src/core/vertexmodel.h [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+    explicit VertexModel( QObject *parent = nullptr );
+    ~VertexModel() override = default;
+
+    //! \copydoc mapSettings
+    void setMapSettings( MapSettings *mapSettings );
+    //! \copydoc mapSettings
+    MapSettings *mapSettings();
+
+    Q_INVOKABLE void setGeometry( const QgsGeometry &geometry, const QgsCoordinateReferenceSystem &crs );
```
Comment (m-kuhn) [inline src/core/vertexmodel.h]: Should the CRS be a property? It somehow feels weird to have the geometry setter different to the getter.

Comment (m-kuhn) [reply src/core/vertexmodel.h]: what I did for something similar recently was this
https://github.com/opengisch/QField/pull/349/files#diff-42599fbe3ce295020d476f85594b8868R125

it basically would not reconstruct the geometry immediately, but only in the next event loop run. If multiple properties are set at once, that will only trigger one expensive recalculation.

But I'd be happy with a simple getter as you propose too.

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-b25cd75bfd3ebc1dd5e14d18c25e2cba03b07288-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-b25cd75bfd3ebc1dd5e14d18c25e2cba03b07288-x86.apk)

**Comment by m-kuhn** [comment]
* ```
   09-03 11:59:22.226 10576 10595 I Qt JAVA : Can't find '/data/app/ch.opengis.qfield-eAA8hV9872TwFGJ4Fz6BLQ==/lib/arm/libqgis_gui.so
   09-03 11:59:22.226 10576 10595 I Qt JAVA : Can't find '/system/lib/libqgis_gui.so
   ```

* It doesn't ask for multimedia permissions yet

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-8fc4ecb273defbe0591b428224282f4afd3ecb76-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-8fc4ecb273defbe0591b428224282f4afd3ecb76-x86.apk)

**Comment by giedrioks26** [comment]
Test apk crashes when opening qfield

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-0f4428b71b664fcac8653e4a2e2dd30c326f702c-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-0f4428b71b664fcac8653e4a2e2dd30c326f702c-x86.apk)

**Comment by m-kuhn** [comment]
packaging works :tada: 
but I don't manage to edit the geometries.
as soon as I click on the "edit geometry" button, the feature form disappears, but then nothing happens.

**Comment by 3nids** [comment]
so do I....I'll check what's going on

**Comment by 3nids** [comment]
works on the computer, but not on the phone

**Comment by giedrioks26** [comment]
> packaging works tada
but I don't manage to edit the geometries.
as soon as I click on the "edit geometry" button, the feature form disappears, but then nothing happens.

same for me

**Comment by Naturtecnica** [comment]
Hi guys, any news about this new skill? When will the new release be on Google Play?
I look forward to testing it.
Thanx, your work's great

**Comment by m-kuhn** [comment]
Soon, just had to do some other cleanups.

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-97774e558626d8391729126c0620657174ef5f21-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-97774e558626d8391729126c0620657174ef5f21-x86.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-3a0cd9116e1ea6242e28d032f6207112c749b128-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-3a0cd9116e1ea6242e28d032f6207112c749b128-x86.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-43bd7700ed3c70b29f2c3ffa9346997069d0a0c6-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-43bd7700ed3c70b29f2c3ffa9346997069d0a0c6-x86.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-40b746f5d91e0511f75da6b2b69ce4ef0e171e7e-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-40b746f5d91e0511f75da6b2b69ce4ef0e171e7e-x86.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-46628493beca20e912b3c584499a7308606d3568-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-46628493beca20e912b3c584499a7308606d3568-x86.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-486da38e53302983c518d6913cbe1fc332650652-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-486da38e53302983c518d6913cbe1fc332650652-x86.apk)

**Comment by m-kuhn** [comment]
Debug hint:

Get a message when a property is recalculated:

```
visible: console.warn( "state: " + stateMachine.state + "Count" + vertexModel.vertexCount ) || ( stateMachine.state === "digitize" && vertexModel.vertexCount > 0 )
```

https://github.com/opengisch/QField/blob/2f2be6d7ea2264058026c5522096f23ddda2234d/src/qml/qgismobileapp.qml#L545

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-d27571225e14a55f79c87529481cd83c889a5723-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-d27571225e14a55f79c87529481cd83c889a5723-x86.apk)

**Comment by giedrioks26** [comment]
> Uploaded test apks for armv7 and x86

Improvement is not working...

**Comment by m-kuhn** [comment]
@giedrioks26 I think @3nids is very well aware that it's still not working and just creating test builds for debug reasons to find out, what's going on ;)

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-5ad7a2dcf9413fc302fa89cae78b550ba10cd116-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-5ad7a2dcf9413fc302fa89cae78b550ba10cd116-x86.apk)

**Comment by m-kuhn** [comment]
It's looking a bit better now. I am able to modify a geometry but saving does not work. The test layer here is reported as multi line string, so maybe that's the reason.

For some reason,  `featureForm.selection.selectedFeature.geometry` returned undefined. The feature was still valid (e.g. id could be accessed) but not the geometry. Still not sure if it's a problem with Q_GADGET in Q_GADET (feature.geometry) and why I had crashes (in `QgsFeature::operator=`) before changing the setGeometry to a property accessor...

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-6d373ec85868d61c494d7422a06b6dfdb7727b93-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-6d373ec85868d61c494d7422a06b6dfdb7727b93-x86.apk)

File: src/core/vertexmodel.h [hunk 7]
```diff
@@ -75,9 +85,25 @@ class VertexModel : public QStandardItemModel
...
+     */
+    QgsCoordinateReferenceSystem crs() const;
+
+    /**
+     * The geometry in layer coordinates
+     */
+    void setGeometry( const QgsGeometry &geometry );
+
+    /**
+     * The geometry in layer coordinates
```
Comment (3nids) [inline src/core/vertexmodel.h]: \copydoc for all 4?

File: src/core/vertexmodel.h [hunk 8]
```diff
@@ -51,6 +51,16 @@ class VertexModel : public QStandardItemModel
     //! determines if one can go to next vertex
     Q_PROPERTY( bool canNextVertex READ canNextVertex NOTIFY canNextVertexChanged )
 
+    /**
+     * The geometry in layer coordinates
```
Comment (3nids) [inline src/core/vertexmodel.h]: add a note that the CRS should be set first?

File: src/core/vertexmodel.h (hunk 7 above)
Comment (m-kuhn) [reply src/core/vertexmodel.h]: I like copydoc. sipify doesn't. That's why I stopped using it. What's your opinion?

Comment (3nids) [reply src/core/vertexmodel.h]: I don't mind, but I have been using it intensively lately.  Just that we do the same.

**Comment by 3nids** [comment]
There is a warning that CRS is not set.
I [added](https://github.com/qgis/QGIS/commit/9c97a341fe5396b3d7ae482d655382237d0ddba6#diff-6b8522a51b843f89703c36e4bab9aa6c) the Crs property on QgsMapLayer on 2018.08.20
but the SDK seems to at [20180815](https://github.com/opengisch/QField/blob/master/.travis.yml#L22) 
might be worth an upgrade....?

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-aeca2abb038261bd008fd72ef23288a49466c696-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-aeca2abb038261bd008fd72ef23288a49466c696-x86.apk)

**Comment by giedrioks26** [comment]
works! I can change the position of the item. It also allows you to save changes. Tested with point, line and polygon.

**Comment by 3nids** [comment]
@m-kuhn @signedav more testing before merging?

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-5589632c7a752b2caef00120ca7fb8a83953b4b1-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-5589632c7a752b2caef00120ca7fb8a83953b4b1-x86.apk)

**Comment by 3nids** [comment]
for me ready to go.
there is just something not perfect, there is a lag after validating the change before the map actually refresh. It would be nice to drop the rubber band only after map is refreshed, but that could be solved later on.

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-ec60ef543c9ba6597709e98b70e72a5246017089-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-348-ec60ef543c9ba6597709e98b70e72a5246017089-x86.apk)

**Pull request merged**